core/impl/src/main/java/com/blazebit/persistence/impl/WindowManager.java

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright 2014 - 2019 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.blazebit.persistence.impl;
+
+import com.blazebit.persistence.WindowBuilder;
+import com.blazebit.persistence.impl.transform.ExpressionModifierVisitor;
+import com.blazebit.persistence.parser.expression.Expression;
+import com.blazebit.persistence.parser.expression.OrderByItem;
+import com.blazebit.persistence.parser.expression.WindowDefinition;
+import com.blazebit.persistence.parser.expression.WindowFrameExclusionType;
+import com.blazebit.persistence.parser.expression.WindowFramePositionType;
+import com.blazebit.persistence.parser.expression.modifier.ExpressionModifier;
+import com.blazebit.persistence.parser.predicate.Predicate;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ *
+ * @author Christian Beikov
+ * @since 1.4.0
+ */
+public class WindowManager<T> extends AbstractManager<ExpressionModifier> {
+
+    private final Map<String, WindowDefinition> windows;
+    private QueryContext queryContext;
+    private WindowBuilder<T> windowBuilder;
+
+    public WindowManager(ResolvingQueryGenerator queryGenerator, ParameterManager parameterManager, SubqueryInitiatorFactory subqueryInitFactory) {
+        super(queryGenerator, parameterManager, subqueryInitFactory);
+        this.windows = new LinkedHashMap<>(0);
+    }
+
+    @Override
+    public void apply(ExpressionModifierVisitor<? super ExpressionModifier> visitor) {
+    }
+
+    public WindowDefinition resolve(WindowDefinition windowDefinition) {
+        if (windowDefinition.getWindowName() == null) {
+            return windowDefinition;
+        }
+
+        WindowDefinition baseWindowDefinition = windows.get(windowDefinition.getWindowName());
+        // See https://www.postgresql.org/docs/current/sql-select.html
+        if (baseWindowDefinition == null) {
+            throw new IllegalArgumentException("There is no window named '" + windowDefinition.getWindowName() + "' registered!");
+        }
+        if (baseWindowDefinition.getFrameMode() != null) {
+            throw new IllegalArgumentException("The base window '" + windowDefinition.getWindowName() + "' is not allowed to specify a frame clause when being reused!");
```

Comment:
The reason being:

> For rather inscrutable reasons, SQL:2008 disallows copying-and-modifying a window definition that has any explicit framing clause.  The error message we gave for this only made sense if the referencing window definition itself contains an explicit framing clause, which it might well not. Moreover, in the context of an OVER clause it’s not exactly obvious that “OVER (windowname)” implies copy-and-modify while “OVER windowname” does not.  This has led to multiple complaints, eg bug #5199 from Iliya Krapchatov.  Change to a hopefully more intelligible error message, and in the case where we have just “OVER (windowname)“, add a HINT suggesting that omitting the parentheses will fix it.  Also improve the related documentation.  Back-patch to all supported branches.

The documentation in the code elaborates:

>  ```
> /*
>   * Per spec, a windowdef that references a previous one copies the
>   * previous partition clause (and mustn't specify its own).  It can
>   * specify its own ordering clause, but only if the previous one had
>   * none.  It always specifies its own frame clause, and the previous
>   * one must not have a frame clause.  Yeah, it's bizarre that each of
>   * these cases works differently, but SQL:2008 says so; see 7.11
>   * <window clause> syntax rule 10 and general rule 1.  The frame
>   * clause rule is especially bizarre because it makes "OVER foo"
>   * different from "OVER (foo)", and requires the latter to throw an
>   * error if foo has a nondefault frame clause.	Well, ours not to
>   * reason why, but we do go out of our way to throw a useful error
>   * message for such cases.
>   */
> ```

So the subtility here is the difference between `OVER x` and `OVER (x)` - the first one executing the window function over window `x` and the latter one using a copied-and-modified version of `x`, which is not allowed for windows that specify a range. So in conclusion: reuse of named windows with range clauses can be allowed just fine, als long as they are rendered without copy operation if not inlined. As we currently inline the windows anyway, we should imo just allow reuse of windows that specify a range clause.

Not having seen the copy-and-modify syntax used to its full extend before, I decided to look into it more, and it seems that its even allowed between different named windows. Obviously with the aforementioned limitations in mind. 

For example:

```pgSQL
SELECT foo.*,
       bar.*,
       row_numbeR() OVER w
FROM (
      VALUES (1) , (2), (3)) AS foo(a),
     (
      VALUES (1), (2), (3)) AS bar(b) window x AS (PARTITION BY a),
                                             w AS (x
                                                   ORDER BY b ROWS BETWEEN unbounded preceding AND CURRENT ROW) ;
```

```
 a | b | row_number 
---+---+------------
 1 | 1 |          1
 1 | 2 |          2
 1 | 3 |          3
 2 | 1 |          1
 2 | 2 |          2
 2 | 3 |          3
 3 | 1 |          1
 3 | 2 |          2
 3 | 3 |          3
```

Above is allowed in postgres (copy-and-modify window definition for another named window)


The below is not:

```pgSQL
SELECT foo.*,
       bar.*,
       row_numbeR() OVER w
FROM (
      VALUES (1) , (2), (3)) AS foo(a),
     (
      VALUES (1), (2), (3)) AS bar(b) window x AS (PARTITION BY a ROWS BETWEEN unbounded preceding AND CURRENT ROW),
                                             w AS (x
                                                   ORDER BY b) ;
```

```
ERROR:  cannot copy window "x" because it has a frame clause
LINE 1: ...ween unbounded preceding and current row ) , w as ( x order ...
```